### PR TITLE
[2] feat(wallets): support eip6963 multi-injected providers

### LINKED ISSUE #3987: [EIP-6963] Multi Injected Provider

### PR DESCRIPTION
**Pull request by shoom3301**
# Summary

Fixes #3987

The implementation is done following the specification: https://eips.ethereum.org/EIPS/eip-6963

## Technical details

1. Added `multiInjectedProvidersAtom` which contains all announced injected providers ([commit](https://github.com/cowprotocol/cowswap/pull/4416/commits/2be5b6d5bd104e8c2ba877b2f9d6ce9adc3b0b01))
2. Added `Eip6963Option` component to display a button to connect Eip6963 provider ([commit](https://github.com/cowprotocol/cowswap/pull/4416/commits/4887f32a2df524dbcbfa725d08115d1d37a744a7#diff-82cf084315959727f558e42e20af603bf0a3bdf00b24be16ac884b7d966bafb4R116))
3. Added `selectedEip6963ProviderAtom` which contains current selected Eip6963 provider. In the next commits it will be renamed to `selectedEip6963ProviderRdnsAtom` ([commit](https://github.com/cowprotocol/cowswap/pull/4416/commits/4887f32a2df524dbcbfa725d08115d1d37a744a7#diff-aebb3e19e3823d5ba6beab9c5724f0ca47417367148900bd3bd0f80e05d971dfR11))
4. Updated `useWalletMetaData()` in order to support Eip6963 providers ([commit](https://github.com/cowprotocol/cowswap/pull/4416/commits/4887f32a2df524dbcbfa725d08115d1d37a744a7#diff-1d24e6d6632f8d98af8580353e3029f209ecb3e482132d5c284a13a393c661bfR73))
5. We used to have `INJECTED` and `INJECTED_WIDGET` connection types and it is very confusing. Because in fact, `INJECTED` is Metamask and `INJECTED_WIDGET` is any injected provider. Since now we don't have a special flow for Metamask and consider it as any other Eip6963 provider I removed `INJECTED` and renamed `INJECTED_WIDGET` -> `INJECTED`. Now we have only `INJECTED` and it covers any injected wallets (Eip6963, window.ethereum, widget) ([commit](https://github.com/cowprotocol/cowswap/pull/4416/commits/7f7493cd1d736a187cc2800bc908445be021cb54))
6. Added a migration to support the change above. The migration resets `selectedWallet` and `selectedWalletBackfilled` values in `localStorage` ([commit](https://github.com/cowprotocol/cowswap/pull/4416/commits/7f7493cd1d736a187cc2800bc908445be021cb54#diff-5949ecab31ba0bc945856ed3ea25758f9c12641bd744fc4efef6da1298455073R8))
7. Enhanced `useEagerlyConnect()` to support Eip6963 providers. We have `selectedWallet` parameter in redux store and `selectedEip6963ProviderRdnsAtom` in jotai atom. Both values are persisted in `localStorage`. Now `useEagerlyConnect()` checks, if there is a `selectedWallet` and `selectedEip6963ProviderRdns` in `localStorage`, then we wait until selected Eip6963 provider is announced and then connect it ([commit](https://github.com/cowprotocol/cowswap/pull/4416/commits/4c85a79fb1ffe7056dbc674c147d9a5e08e2557f))
8. We have separate connection buttons in the wallets list for Coinbase and Trust. But when there are injected providers for those wallets it doesn't make sense to display them. So I added conditions for the buttons displaying. P.s. Trust wallet icon will be deleted in the next PR ([commit](https://github.com/cowprotocol/cowswap/pull/4416/commits/af95b2cbecbf286ba6a79ffffd299fc083335eb9))

## Functional changes
1. When there is an inejcted Conibase wallet, then we don't display a special button for it. Otherwise, we display the special button which opens a modal with QR.
2. Please ignore special Trust wallet icon, it will be deleted in [the next PR](https://github.com/cowprotocol/cowswap/pull/4420). So, the wallet should work as regular EIP6963 provider.
3. #3541 case was affected. I've tested it and it was good, but anyway, please take into account.

image

# To Test

Please, do tests in the last PR of the waterfall: https://github.com/cowprotocol/cowswap/pull/4420

### CODE REVIEW

File: libs/wallet/src/web3-react/connectors/Injected/index.tsx [hunk 1]
```diff
@@ -234,7 +214,18 @@ export class InjectedWallet extends Connector {
         Failed to get account with eth_requestAccounts method
         Trying with eth_accounts method
       `)
+
       return provider.request({ method: 'eth_accounts' }) as Promise<string[]>
     }
   }
 }
+
+function parseChainId(chainId: string | number): number {
```
Comment (shoom3301) [inline libs/wallet/src/web3-react/connectors/Injected/index.tsx:228]: Just moved it from the file beginning

File: libs/wallet/src/web3-react/Web3Provider/hooks/useEagerlyConnect.ts [hunk 2]
```diff
@@ -24,19 +28,33 @@ async function connect(connector: Connector) {
...
-    if (window.top !== window.self) {
+    if (isIframe) {
       connect(gnosisSafeConnection.connector)
     }
 
     connect(networkConnection.connector)
 
     if (selectedWallet) {
+      const connection = getWeb3ReactConnection(selectedWallet)
+      const cachedProviderRdns = jotaiStore.get(selectedEip6963ProviderRdnsAtom)
```
Comment (shoom3301) [inline libs/wallet/src/web3-react/Web3Provider/hooks/useEagerlyConnect.ts:50]: I could use `useAtomValue(selectedEip6963ProviderRdnsAtom)` instead, but for some reason it doesn't work as expected. With `useAtomValue` it doesn't have enough time to fullfill data from `localStorage`.
Anyway, this `useEffect()` intentionally doesn't have dependencies, so it's even better  if we use `jotaiStore.get()` here.

File: libs/wallet/src/web3-react/connection/injectedOptions.tsx [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    <ConnectWalletOption
+      {...options}
+      isActive={isActive}
+      header={getConnectionName(ConnectionType.INJECTED)}
+      onClick={() => tryActivation(injectedWalletConnection.connector)}
+    />
+  )
+}
+
+interface Eip6963OptionProps {
```
Comment (shoom3301) [inline libs/wallet/src/web3-react/connection/injectedOptions.tsx:72]: All code above is not new, it used to be in `injected.tsx`

**Comment by alfetopito** [comment]
Have not reviewed yet, just based on the PR description, the item `6`:

> Added a migration to support the change above. The migration resets selectedWallet and selectedWalletBackfilled values in localStorage ([commit](https://github.com/cowprotocol/cowswap/pull/4416/commits/7f7493cd1d736a187cc2800bc908445be021cb54#diff-5949ecab31ba0bc945856ed3ea25758f9c12641bd744fc4efef6da1298455073R8))

Won't the [hotfix to clear the local storage](https://cowservices.slack.com/archives/C0361CDG8GP/p1715590791026739) make this migration unreachable?
My understanding is that the storage clearing happens on emergency.js, before the app is loaded.
Then, when we look for the migration, there's no old state to clear as it was removed, no?

File: libs/wallet/src/api/state/multiInjectedProvidersAtom.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+import { getJotaiIsolatedStorage, jotaiStore } from '@cowprotocol/core'
+
+import { EIP6963AnnounceProviderEvent, EIP6963ProviderDetail } from '../eip6963-types'
+
+export const multiInjectedProvidersAtom = atom<EIP6963ProviderDetail[]>([])
+
+// RDNS of the selected EIP-6963 provider
+export const selectedEip6963ProviderRdnsAtom = atomWithStorage<string | null>(
+  'selectedEip6963ProviderAtom:v1',
```
Comment (alfetopito) [inline libs/wallet/src/api/state/multiInjectedProvidersAtom.ts:12]: Is this the first? Start with 0?

```suggestion
  'selectedEip6963ProviderAtom:v0,
```

File: libs/wallet/src/web3-react/connectors/Injected/index.tsx [hunk 5]
```diff
@@ -141,7 +119,7 @@ export class InjectedWallet extends Connector {
 
   // Based on https://github.com/Uniswap/web3-react/blob/de97c00c378b7909dfbd8a06558ed12e1f796caa/packages/metamask/src/index.ts#L54 with some changes
   private async isomorphicInitialize(): Promise<void> {
-    // Mod: we have a custom method to detect a Injected provider based on passed keywords array
+    // Mod: we have a custom method to detect Injected provider based on passed keywords array
```
Comment (alfetopito) [inline libs/wallet/src/web3-react/connectors/Injected/index.tsx:122]: Nitpick: remove `Mod:` comments

```suggestion
```

**Review by alfetopito — APPROVED** [review]

File: libs/wallet/src/web3-react/connectors/Injected/index.tsx (hunk 5 above)
Comment (shoom3301) [reply libs/wallet/src/web3-react/connectors/Injected/index.tsx:122]: Fixed https://github.com/cowprotocol/cowswap/pull/4420/commits/c98d027b744ea771203d50a39eda74cb1f7a0215

File: libs/wallet/src/api/state/multiInjectedProvidersAtom.ts (hunk 4 above)
Comment (shoom3301) [reply libs/wallet/src/api/state/multiInjectedProvidersAtom.ts:12]: Fixed https://github.com/cowprotocol/cowswap/pull/4420/commits/c98d027b744ea771203d50a39eda74cb1f7a0215

**Review by elena-zh — COMMENTED** [review]
See https://github.com/cowprotocol/cowswap/pull/4420#pullrequestreview-2058732832

**Comment by shoom3301** [comment]
> Have not reviewed yet, just based on the PR description, the item `6`:
> 
> > Added a migration to support the change above. The migration resets selectedWallet and selectedWalletBackfilled values in localStorage ([commit](https://github.com/cowprotocol/cowswap/pull/4416/commits/7f7493cd1d736a187cc2800bc908445be021cb54#diff-5949ecab31ba0bc945856ed3ea25758f9c12641bd744fc4efef6da1298455073R8))
> 
> Won't the [hotfix to clear the local storage](https://cowservices.slack.com/archives/C0361CDG8GP/p1715590791026739) make this migration unreachable? My understanding is that the storage clearing happens on emergency.js, before the app is loaded. Then, when we look for the migration, there's no old state to clear as it was removed, no?

The fix in `emergency.js` only fix localStorage keys with versions.
The migration in this PR is about `redux_localstorage_simple_user` localStorage key.

**Pull request merged**